pkg/storage/tsdb/single_file_index.go

```diff
@@ -0,0 +1,114 @@
+package tsdb
+
+import (
+	"context"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/pkg/storage/tsdb/index"
+)
+
+// nolint
+type TSDBIndex struct {
+	reader IndexReader
+}
+
+func NewTSDBIndex(reader IndexReader) *TSDBIndex {
+	return &TSDBIndex{
+		reader: reader,
+	}
+}
+
+func (i *TSDBIndex) GetChunkRefs(_ context.Context, userID string, from, through model.Time, matchers ...*labels.Matcher) ([]ChunkRef, error) {
+
+	p, err := PostingsForMatchers(i.reader, matchers...)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		res  []ChunkRef // TODO(owen-d): pool, reduce allocs
+		ls   labels.Labels
+		chks []index.ChunkMeta
+	)
+
+	for p.Next() {
+		if err := i.reader.Series(p.At(), &ls, &chks); err != nil {
+			return nil, err
+		}
+
+		// cache hash calculation across chunks of the same series
+		// TODO(owen-d): Should we store this in the index? in an in-mem cache?
+		var hash uint64
+
+		// TODO(owen-d): use logarithmic approach
+		for _, chk := range chks {
+
+			// current chunk is outside the range of this request
+			if chk.From() > through || chk.Through() < from {
+				continue
+			}
+
+			if hash == 0 {
+				hash = ls.Hash()
```

Comment:
are we not saving the hash somewhere ?

Reply:
Not yet, no. I was looking at this yesterday, but I haven't set it up yet.